File: diffmerge-1/py/diffmerge/patch.py
"""Patch functionality for the diffmerge library."""

import re
from typing import Optional, Dict, Any, List
from .types import (
    ApplyResult, ParsedPatch, PatchHunk, PatchLine,
    PatchOptions, ParseError
)
from .diff import diff_lines
from .utils import split_lines


def create_patch(old: str, new: str, options: Optional[Dict[str, Any]] = None) -> str:
    """Generate unified diff format patch.

    Args:
        old: Original content
        new: New content
        options: Optional dict with keys:
            - old_file: Name of old file (default: "a")
            - new_file: Name of new file (default: "b")
            - context_lines: Context lines (default: 3)

    Returns:
        Unified diff format string
    """
    opts = options or {}
    old_file = opts.get("old_file", "a")
    new_file = opts.get("new_file", "b")
    context_lines = opts.get("context_lines", 3)

    diff_result = diff_lines(old, new, {"context_lines": context_lines})
    hunks = diff_result["hunks"]

    if not hunks:
        # Both empty
        return f"--- {old_file}\n+++ {new_file}\n"

    # Check if there are any changes
    has_changes = any(h["op"] != "equal" for h in hunks)
    if not has_changes:
        return f"--- {old_file}\n+++ {new_file}\n"

    # Group hunks into unified diff hunks with context
    result_lines = [f"--- {old_file}", f"+++ {new_file}"]

    # Find change regions and expand with context
    change_indices = [i for i, h in enumerate(hunks) if h["op"] != "equal"]

    if not change_indices:
        return f"--- {old_file}\n+++ {new_file}\n"

    # Group changes that are close enough to share context
    groups = []
    current_group = [change_indices[0]]

    for i in range(1, len(change_indices)):
        # If this change is within 2*context_lines of the previous, group them
        if change_indices[i] - change_indices[i-1] <= 2 * context_lines:
            current_group.append(change_indices[i])
        else:
            groups.append(current_group)
            current_group = [change_indices[i]]
    groups.append(current_group)

    # Generate a unified hunk for each group
    for group in groups:
        # Calculate range with context
        start = max(0, group[0] - context_lines)
        end = min(len(hunks), group[-1] + context_lines + 1)

        # Calculate line numbers for the hunk header
        # old_start: starting line in old file
        # new_start: starting line in new file
        old_start = 1
        new_start = 1

        for i in range(start):
            if hunks[i]["op"] == "equal":
                old_start += 1
                new_start += 1
            elif hunks[i]["op"] == "delete":
                old_start += 1
            else:  # insert
                new_start += 1

        # Count lines in this hunk
        old_count = 0
        new_count = 0
        hunk_lines = []

        for i in range(start, end):
            h = hunks[i]
            content = h["content"]
            # Remove trailing newline for display, we'll add it back
            if content.endswith("\n"):
                line_content = content[:-1]
            elif content.endswith("\r\n"):
                line_content = content[:-2]
            elif content.endswith("\r"):
                line_content = content[:-1]
            else:
                line_content = content

            if h["op"] == "equal":
                hunk_lines.append(f" {line_content}")
                old_count += 1
                new_count += 1
            elif h["op"] == "delete":
                hunk_lines.append(f"-{line_content}")
                old_count += 1
            else:  # insert
                hunk_lines.append(f"+{line_content}")
                new_count += 1

        # Generate hunk header
        if old_count == 0:
            old_start = max(old_start - 1, 0)  # Point to line before insertion
        if new_count == 0:
            new_start = max(new_start - 1, 0)

        header = f"@@ -{old_start},{old_count} +{new_start},{new_count} @@"
        result_lines.append(header)
        result_lines.extend(hunk_lines)

    return "\n".join(result_lines) + "\n"


def parse_patch(patch: str) -> ParsedPatch:
    """Parse unified diff format into structured data.

    Args:
        patch: Unified diff format string

    Returns:
        ParsedPatch with file names and hunks

    Raises:
        ParseError: If the patch format is invalid
    """
    lines = patch.split("\n")

    old_file = ""
    new_file = ""
    hunks: List[PatchHunk] = []

    i = 0

    # Find file headers
    while i < len(lines):
        line = lines[i]
        if line.startswith("---"):
            old_file = line[4:].strip()
            i += 1
            break
        i += 1
    else:
        raise ParseError("No --- header found in patch")

    while i < len(lines):
        line = lines[i]
        if line.startswith("+++"):
            new_file = line[4:].strip()
            i += 1
            break
        i += 1
    else:
        raise ParseError("No +++ header found in patch")

    # Parse hunks
    hunk_header_re = re.compile(r"@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@")

    while i < len(lines):
        line = lines[i]
        if not line:
            i += 1
            continue

        match = hunk_header_re.match(line)
        if match:
            old_start = int(match.group(1))
            old_count = int(match.group(2)) if match.group(2) else 1
            new_start = int(match.group(3))
            new_count = int(match.group(4)) if match.group(4) else 1

            hunk_lines: List[PatchLine] = []
            i += 1

            # Parse lines in this hunk
            expected_old = old_count
            expected_new = new_count

            while i < len(lines) and (expected_old > 0 or expected_new > 0):
                if i >= len(lines):
                    break
                line = lines[i]
                if line.startswith("@@") or line.startswith("---") or line.startswith("+++"):
                    break

                if line.startswith(" "):
                    hunk_lines.append({"op": " ", "content": line[1:]})
                    expected_old -= 1
                    expected_new -= 1
                elif line.startswith("-"):
                    hunk_lines.append({"op": "-", "content": line[1:]})
                    expected_old -= 1
                elif line.startswith("+"):
                    hunk_lines.append({"op": "+", "content": line[1:]})
                    expected_new -= 1
                elif line == "":
                    # Empty context line
                    hunk_lines.append({"op": " ", "content": ""})
                    expected_old -= 1
                    expected_new -= 1
                else:
                    # Might be "\ No newline at end of file" or other marker
                    pass
                i += 1

            hunks.append({
                "old_start": old_start,
                "old_count": old_count,
                "new_start": new_start,
                "new_count": new_count,
                "lines": hunk_lines,
            })
        else:
            i += 1

    return {
        "old_file": old_file,
        "new_file": new_file,
        "hunks": hunks,
    }


def apply_patch(content: str, patch: str) -> ApplyResult:
    """Apply a unified diff patch to content.

    Args:
        content: Original content
        patch: Unified diff format patch

    Returns:
        ApplyResult with content and statistics
    """
    errors: List[str] = []

    try:
        parsed = parse_patch(patch)
    except ParseError as e:
        return {
            "content": content,
            "success": False,
            "hunks_applied": 0,
            "hunks_failed": 1,
            "errors": [str(e)],
        }

    # If no hunks, patch is a no-op
    if not parsed["hunks"]:
        return {
            "content": content,
            "success": True,
            "hunks_applied": 0,
            "hunks_failed": 0,
            "errors": [],
        }

    lines = split_lines(content)
    # Normalize lines to not include newlines for comparison
    content_lines = []
    for line in lines:
        if line.endswith("\r\n"):
            content_lines.append(line[:-2])
        elif line.endswith("\n") or line.endswith("\r"):
            content_lines.append(line[:-1])
        else:
            content_lines.append(line)

    hunks_applied = 0
    hunks_failed = 0

    # Apply hunks in reverse order to not mess up line numbers
    offset = 0  # Track cumulative offset from applied hunks

    for hunk in parsed["hunks"]:
        # Find where to apply this hunk
        old_start = hunk["old_start"] - 1 + offset  # Convert to 0-indexed
        old_count = hunk["old_count"]

        # Get expected old lines from hunk
        expected_lines = [pl["content"] for pl in hunk["lines"] if pl["op"] in (" ", "-")]

        # Try exact match first
        actual_start = old_start
        matched = False

        # Check if expected lines match at the expected position
        if actual_start >= 0 and actual_start + len(expected_lines) <= len(content_lines):
            match = True
            for j, expected in enumerate(expected_lines):
                if content_lines[actual_start + j] != expected:
                    match = False
                    break
            if match:
                matched = True

        # Try fuzzy matching with offset
        if not matched:
            for fuzz in range(1, 10):
                for direction in [1, -1]:
                    test_start = old_start + fuzz * direction
                    if test_start >= 0 and test_start + len(expected_lines) <= len(content_lines):
                        match = True
                        for j, expected in enumerate(expected_lines):
                            if content_lines[test_start + j] != expected:
                                match = False
                                break
                        if match:
                            actual_start = test_start
                            matched = True
                            break
                if matched:
                    break

        if not matched:
            hunks_failed += 1
            errors.append(f"Hunk at line {hunk['old_start']} failed to apply")
            continue

        # Apply the hunk
        new_lines = []
        for pl in hunk["lines"]:
            if pl["op"] == " " or pl["op"] == "+":
                new_lines.append(pl["content"])

        # Replace old lines with new lines
        content_lines = (
            content_lines[:actual_start] +
            new_lines +
            content_lines[actual_start + len(expected_lines):]
        )

        # Update offset
        offset += len(new_lines) - len(expected_lines)
        hunks_applied += 1

    # Reconstruct content
    result_content = "\n".join(content_lines)
    if content_lines and (content.endswith("\n") or content == ""):
        result_content += "\n"

    success = hunks_failed == 0

    return {
        "content": result_content,
        "success": success,
        "hunks_applied": hunks_applied,
        "hunks_failed": hunks_failed,
        "errors": errors,
    }


def reverse_patch(patch: str) -> str:
    """Reverse a patch (swap additions and deletions).

    Args:
        patch: Unified diff format patch

    Returns:
        Reversed unified diff format patch
    """
    lines = patch.split("\n")
    result = []

    hunk_header_re = re.compile(r"@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@(.*)")

    old_file = None
    new_file = None

    i = 0
    while i < len(lines):
        line = lines[i]
        if line.startswith("--- "):
            # Store old file name, but don't output yet
            old_file = line[4:]
            i += 1
            continue
        elif line.startswith("+++ "):
            # Store new file name, now output both swapped
            new_file = line[4:]
            # Output --- with new file name and +++ with old file name
            result.append("--- " + new_file)
            result.append("+++ " + old_file)
            i += 1
            continue
        elif line.startswith("@@"):
            # Swap line counts in hunk header
            match = hunk_header_re.match(line)
            if match:
                old_start = match.group(1)
                old_count = match.group(2) if match.group(2) else "1"
                new_start = match.group(3)
                new_count = match.group(4) if match.group(4) else "1"
                suffix = match.group(5) if match.group(5) else ""
                result.append(f"@@ -{new_start},{new_count} +{old_start},{old_count} @@{suffix}")
            else:
                result.append(line)
        elif line.startswith("-"):
            # Change deletion to addition
            result.append("+" + line[1:])
        elif line.startswith("+"):
            # Change addition to deletion
            result.append("-" + line[1:])
        else:
            result.append(line)
        i += 1

    return "\n".join(result)

File: diffmerge-2/go/dm/patch.go
package dm

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// CreatePatch generates a unified diff format patch
func CreatePatch(old, new string, options PatchOptions) string {
	if options.OldFile == "" {
		options.OldFile = "a"
	}
	if options.NewFile == "" {
		options.NewFile = "b"
	}
	if options.ContextLines == 0 {
		options.ContextLines = 3
	}

	oldLines := SplitLines(old)
	newLines := SplitLines(new)

	// Compute the diff
	diffResult := DiffLines(old, new, DiffOptions{ContextLines: options.ContextLines})

	// If no changes, return minimal patch
	if diffResult.Stats.Additions == 0 && diffResult.Stats.Deletions == 0 {
		return fmt.Sprintf("--- %s\n+++ %s\n", options.OldFile, options.NewFile)
	}

	// Group hunks into patch hunks
	patchHunks := groupIntoPatchHunks(diffResult.Hunks, oldLines, newLines, options.ContextLines)

	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("--- %s\n", options.OldFile))
	sb.WriteString(fmt.Sprintf("+++ %s\n", options.NewFile))

	for _, hunk := range patchHunks {
		// Write hunk header
		sb.WriteString(fmt.Sprintf("@@ -%d,%d +%d,%d @@\n",
			hunk.OldStart, hunk.OldCount, hunk.NewStart, hunk.NewCount))

		// Write hunk lines
		for _, line := range hunk.Lines {
			sb.WriteString(line.Op)
			sb.WriteString(line.Content)
			sb.WriteString("\n")
		}
	}

	return sb.String()
}

// groupIntoPatchHunks groups diff hunks into patch hunks with context
func groupIntoPatchHunks(hunks []DiffHunk, oldLines, newLines []string, contextLines int) []PatchHunk {
	if len(hunks) == 0 {
		return nil
	}

	// Find all change regions
	type changeRegion struct {
		startIdx int
		endIdx   int
	}

	var regions []changeRegion
	inChange := false
	startIdx := 0

	for i, h := range hunks {
		if h.Op != OpEqual {
			if !inChange {
				inChange = true
				startIdx = i
			}
		} else {
			if inChange {
				regions = append(regions, changeRegion{startIdx, i})
				inChange = false
			}
		}
	}
	if inChange {
		regions = append(regions, changeRegion{startIdx, len(hunks)})
	}

	if len(regions) == 0 {
		return nil
	}

	// Merge regions that are close together
	var mergedRegions []changeRegion
	current := regions[0]

	for i := 1; i < len(regions); i++ {
		// Check if regions should be merged (gap <= 2*contextLines)
		gap := regions[i].startIdx - current.endIdx
		if gap <= 2*contextLines {
			current.endIdx = regions[i].endIdx
		} else {
			mergedRegions = append(mergedRegions, current)
			current = regions[i]
		}
	}
	mergedRegions = append(mergedRegions, current)

	// Create patch hunks from merged regions
	var patchHunks []PatchHunk

	for _, region := range mergedRegions {
		var lines []PatchLine
		oldStart := 0
		newStart := 0
		oldCount := 0
		newCount := 0

		// Find the actual line numbers
		foundStart := false
		for i := max(0, region.startIdx-contextLines); i < min(len(hunks), region.endIdx+contextLines); i++ {
			h := hunks[i]

			if !foundStart {
				if h.OldStart > 0 {
					oldStart = h.OldStart
				}
				if h.NewStart > 0 {
					newStart = h.NewStart
				}
				if oldStart > 0 || newStart > 0 {
					foundStart = true
				}
			}

			switch h.Op {
			case OpEqual:
				lines = append(lines, PatchLine{Op: " ", Content: h.Content})
				oldCount++
				newCount++
			case OpDelete:
				lines = append(lines, PatchLine{Op: "-", Content: h.Content})
				oldCount++
			case OpInsert:
				lines = append(lines, PatchLine{Op: "+", Content: h.Content})
				newCount++
			}
		}

		if oldStart == 0 {
			oldStart = 1
		}
		if newStart == 0 {
			newStart = 1
		}

		// Adjust start for context
		contextBefore := 0
		for i := region.startIdx - 1; i >= 0 && contextBefore < contextLines; i-- {
			if hunks[i].Op == OpEqual {
				contextBefore++
			}
		}

		if len(lines) > 0 {
			patchHunks = append(patchHunks, PatchHunk{
				OldStart: max(1, oldStart-contextBefore),
				OldCount: oldCount,
				NewStart: max(1, newStart-contextBefore),
				NewCount: newCount,
				Lines:    lines,
			})
		}
	}

	return patchHunks
}

// ParsePatch parses unified diff format into structured data
func ParsePatch(patch string) (ParsedPatch, error) {
	lines := strings.Split(patch, "\n")

	result := ParsedPatch{
		OldFile: "",
		NewFile: "",
		Hunks:   []PatchHunk{},
	}

	hunkHeaderRe := regexp.MustCompile(`^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@`)

	var currentHunk *PatchHunk

	for _, line := range lines {
		if strings.HasPrefix(line, "--- ") {
			result.OldFile = strings.TrimPrefix(line, "--- ")
		} else if strings.HasPrefix(line, "+++ ") {
			result.NewFile = strings.TrimPrefix(line, "+++ ")
		} else if matches := hunkHeaderRe.FindStringSubmatch(line); matches != nil {
			if currentHunk != nil {
				result.Hunks = append(result.Hunks, *currentHunk)
			}

			oldStart, _ := strconv.Atoi(matches[1])
			oldCount := 1
			if matches[2] != "" {
				oldCount, _ = strconv.Atoi(matches[2])
			}
			newStart, _ := strconv.Atoi(matches[3])
			newCount := 1
			if matches[4] != "" {
				newCount, _ = strconv.Atoi(matches[4])
			}

			currentHunk = &PatchHunk{
				OldStart: oldStart,
				OldCount: oldCount,
				NewStart: newStart,
				NewCount: newCount,
				Lines:    []PatchLine{},
			}
		} else if currentHunk != nil && len(line) > 0 {
			op := string(line[0])
			content := ""
			if len(line) > 1 {
				content = line[1:]
			}

			if op == " " || op == "+" || op == "-" {
				currentHunk.Lines = append(currentHunk.Lines, PatchLine{
					Op:      op,
					Content: content,
				})
			}
		}
	}

	if currentHunk != nil {
		result.Hunks = append(result.Hunks, *currentHunk)
	}

	if result.OldFile == "" && result.NewFile == "" && len(result.Hunks) == 0 {
		return result, fmt.Errorf("invalid patch format")
	}

	return result, nil
}

// ApplyPatch applies a unified diff patch to content
func ApplyPatch(content, patch string) ApplyResult {
	parsed, err := ParsePatch(patch)
	if err != nil {
		return ApplyResult{
			Content:      content,
			Success:      false,
			HunksApplied: 0,
			HunksFailed:  0,
			Errors:       []string{err.Error()},
		}
	}

	// If no hunks, it's a no-op patch
	if len(parsed.Hunks) == 0 {
		return ApplyResult{
			Content:      content,
			Success:      true,
			HunksApplied: 0,
			HunksFailed:  0,
			Errors:       []string{},
		}
	}

	lines := SplitLines(content)
	hadTrailingNewline := strings.HasSuffix(content, "\n")

	result := make([]string, 0, len(lines))
	lineIdx := 0
	hunksApplied := 0
	hunksFailed := 0
	errors := []string{}

	for _, hunk := range parsed.Hunks {
		targetLine := hunk.OldStart - 1 // Convert to 0-indexed

		// Copy lines before this hunk
		for lineIdx < targetLine && lineIdx < len(lines) {
			result = append(result, lines[lineIdx])
			lineIdx++
		}

		// Verify and apply hunk
		success := true
		expectedIdx := 0

		// First verify all context and delete lines match
		for _, patchLine := range hunk.Lines {
			if patchLine.Op == " " || patchLine.Op == "-" {
				checkIdx := targetLine + expectedIdx
				if checkIdx >= len(lines) || lines[checkIdx] != patchLine.Content {
					success = false
					break
				}
				expectedIdx++
			}
		}

		if success {
			// Apply the hunk
			for _, patchLine := range hunk.Lines {
				switch patchLine.Op {
				case " ":
					result = append(result, patchLine.Content)
					lineIdx++
				case "-":
					lineIdx++
				case "+":
					result = append(result, patchLine.Content)
				}
			}
			hunksApplied++
		} else {
			// Hunk failed - try with offset
			applied := tryApplyWithOffset(lines, lineIdx, hunk, &result)
			if applied {
				// Update lineIdx based on what was consumed
				for _, patchLine := range hunk.Lines {
					if patchLine.Op == " " || patchLine.Op == "-" {
						lineIdx++
					}
				}
				hunksApplied++
			} else {
				hunksFailed++
				errors = append(errors, fmt.Sprintf("Hunk at line %d failed to apply", hunk.OldStart))
			}
		}
	}

	// Copy remaining lines
	for lineIdx < len(lines) {
		result = append(result, lines[lineIdx])
		lineIdx++
	}

	// Reconstruct content
	var finalContent string
	if len(result) > 0 {
		finalContent = strings.Join(result, "\n")
		if hadTrailingNewline || (len(parsed.Hunks) > 0) {
			finalContent += "\n"
		}
	}

	return ApplyResult{
		Content:      finalContent,
		Success:      hunksFailed == 0,
		HunksApplied: hunksApplied,
		HunksFailed:  hunksFailed,
		Errors:       errors,
	}
}

// tryApplyWithOffset attempts to apply a hunk with fuzzy matching
func tryApplyWithOffset(lines []string, startIdx int, hunk PatchHunk, result *[]string) bool {
	// Try different offsets
	for offset := -3; offset <= 3; offset++ {
		targetLine := startIdx + offset
		if targetLine < 0 {
			continue
		}

		// Verify hunk at this offset
		success := true
		checkIdx := targetLine
		for _, patchLine := range hunk.Lines {
			if patchLine.Op == " " || patchLine.Op == "-" {
				if checkIdx >= len(lines) || lines[checkIdx] != patchLine.Content {
					success = false
					break
				}
				checkIdx++
			}
		}

		if success {
			// Apply at this offset
			for _, patchLine := range hunk.Lines {
				switch patchLine.Op {
				case " ":
					*result = append(*result, patchLine.Content)
				case "-":
					// Skip this line
				case "+":
					*result = append(*result, patchLine.Content)
				}
			}
			return true
		}
	}

	return false
}

// ReversePatch reverses a patch (swap additions and deletions)
func ReversePatch(patch string) string {
	lines := strings.Split(patch, "\n")
	var result []string

	hunkHeaderRe := regexp.MustCompile(`^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@(.*)$`)

	for _, line := range lines {
		if strings.HasPrefix(line, "--- ") {
			result = append(result, "+++ "+strings.TrimPrefix(line, "--- "))
		} else if strings.HasPrefix(line, "+++ ") {
			result = append(result, "--- "+strings.TrimPrefix(line, "+++ "))
		} else if matches := hunkHeaderRe.FindStringSubmatch(line); matches != nil {
			// Swap old and new in hunk header
			oldStart := matches[1]
			oldCount := matches[2]
			if oldCount == "" {
				oldCount = "1"
			}
			newStart := matches[3]
			newCount := matches[4]
			if newCount == "" {
				newCount = "1"
			}
			suffix := matches[5]

			result = append(result, fmt.Sprintf("@@ -%s,%s +%s,%s @@%s",
				newStart, newCount, oldStart, oldCount, suffix))
		} else if len(line) > 0 {
			switch line[0] {
			case '+':
				result = append(result, "-"+line[1:])
			case '-':
				result = append(result, "+"+line[1:])
			default:
				result = append(result, line)
			}
		} else {
			result = append(result, line)
		}
	}

	return strings.Join(result, "\n")
}
